### DISCUSSION TITLE
end based argument continuation

### DISCUSSION THREAD

**Discussion by amitu** [fastn-stack/fastn#388] [Ideas & RFCs]
<details><summary>Status: In Progress</summary>
Owner: @Arpita-Jaiswal 
Started On: 1st September

### TODO

- [x] [P1 Parser](https://github.com/FifthTry/ftd/discussions/390/)
- [x] [DI](https://github.com/FifthTry/ftd/discussions/417)
- [x] [AST](https://github.com/FifthTry/ftd/discussions/413)
- [ ] Interpreter
- [ ] Executor
- [ ] UI
- [ ] Create FTD/FPM release
- [ ] Fix all the ftd files in all fpm repos

#### Common

- padding
- padding-left
- padding-right
- padding-top
- padding-bottom
- padding-horizontal
- padding-vertical
- border-width
- border-radius
- border-top-left-radius
- border-top-right-radius
- border-bottom-left-radius
- border-bottom-right-radius

### Container

- spacing
- wrap

### 27th October

- [Basic event working ✌️, check 7-events.ftd](https://github.com/ftd-lang/ftd/commit/58b1ebf5952613195bd711fb1e1515bcda456936)

### 26th October

- [Change the value on event](https://github.com/ftd-lang/ftd/commit/6c38db17617adae02385e830bb5cf4e2d851f5b7)
- [Added data in html](https://github.com/ftd-lang/ftd/commit/3ed3c08f28e35ea338e99de126274e36618f5881)
- [Added functions in html](https://github.com/ftd-lang/ftd/commit/a97102e2e8dfc6ec64e42eba16358131ea54d8e6)
- [Added events read in html](https://github.com/ftd-lang/ftd/commit/d5aa013d91e0579b1bd8d603e2cb277308ae0425)

### 21st October

- [Added ftd_v2 in examples](https://github.com/ftd-lang/ftd/commit/bc131bfd0c4e0ecc0e392301e1c69dc4f3299778)

### 11th October

- [Recusive loop in html state](https://github.com/ftd-lang/ftd/commit/ebe91e1d2ba8755cc9927b02f698c5ffea840bd2)
- [Added recursive variable in interpreter test](https://github.com/ftd-lang/ftd/commit/a6de08459dd8d192fed210cbf61f82a3d6d33167)
- [Added recursion in component interpreter](https://github.com/ftd-lang/ftd/commit/08da5c8ee51e4825538def95509b98f15e30a2bc)
- [Added recursion in record interpreter](https://github.com/ftd-lang/ftd/commit/0c93b272cbfa90d75b17b5cd62b84f7cdba1893a)

### 10th October

- [Added iteration in Component in executor](https://github.com/ftd-lang/ftd/commit/52977bf66d1a7928ec0a0788804d6c0bc448a5b3)
- [Added iteration in Component in interpreter, wip in executor](https://github.com/ftd-lang/ftd/commit/b7e11673e5bb6def04ed02a808a0e63154615e98)

### 9th October

- [Refactor: Added ExecuteDoc::execute_simple_component](https://github.com/ftd-lang/ftd/commit/4e51c9a647d1464d4245e0aff10bcfd6685f8920)
- [Added data-id in html node](https://github.com/ftd-lang/ftd/commit/66f33fb49e412b8e3ee35af5ff84a121d5fae8c4)
- [Added component definition and its invocation in interpreter](https://github.com/ftd-lang/ftd/commit/365b81379c108b8b06c5ba749be538d62b90031e)
- [Added insert_local_variable in executor tdoc](https://github.com/ftd-lang/ftd/commit/16fe8c2d2158457581d1ae4ca0c8b5fb2e56b0f5)

### 8th October

- Tests [link 1](https://github.com/ftd-lang/ftd/commit/eb80ffba49d5f938d1e5d48ab7adf01ba584f553) [link 2](https://github.com/ftd-lang/ftd/commit/bf4f829440847a320d6c64d9c5279b7571361615)

### 7th October

- [Added html tests](https://github.com/ftd-lang/ftd/commit/e8211fc1408edddb67965f02b7ad0fc8e657dd6e)

### 6th October

- [Fix html test and added ftd.column in executor](https://github.com/ftd-lang/ftd/commit/2727cc4e452f1c1988d17db4a38627de77c3acf6)

### 5th October

- [Added `ftd.row` in executor](https://github.com/ftd-lang/ftd/commit/5d7dae5c32d37f1d4ba2f3ef5a3e1f667d947c14)
- [Node and html added](https://github.com/ftd-lang/ftd/commit/d3f512f805e3a591dcc628a2824cbfa617c7d511)
- [Executor added test](https://github.com/ftd-lang/ftd/commit/2473dc4c3f4c39b1a6be2c5e473f631c39a1b616)
- [Executor initial setup](https://github.com/ftd-lang/ftd/commit/fcbc988cc14d9cc8e7a6d4f0031f83cd1aa8678a)

### 4th October

- [Added intepreter component-invocation test](https://github.com/ftd-lang/ftd/commit/3761efec88232f9673e80c658a016d97285336aa)

### 3rd October

- [Interpreter ComponentDefinition done](https://github.com/ftd-lang/ftd/commit/c7fcbc9011e815ada57be9d1dea6466578dfb980)
- [Interpreter Component done and added to interpreter state](https://github.com/ftd-lang/ftd/commit/dd903190cf1d40d6cf2429845ddc63c01cb45284)

### 2nd October

- [x] [Set variable function in interpreter](https://github.com/ftd-lang/ftd/commit/f39107fba38e9efa4d1ecb1ebb3917a01efd9d93)
- [ ] [Interpreter ComponentDefinition](https://github.com/ftd-lang/ftd/commit/ca7239f6ffc2340b3fad2722f49b9bebe1df31c7)

### 29th September

- [Added variable_definition to interpreter](https://github.com/ftd-lang/ftd/commit/70d6ba71526981d9b04cbcf1c6e0a590138fc951)
- [Added variable_invocation to interpreter](https://github.com/ftd-lang/ftd/commit/0db1ef2ec799af5cec1ec71ae149c4a07f68ee54)
- [Added expression in interpreter](https://github.com/ftd-lang/ftd/commit/bf30fbacfc334e2d31b23a5a3edaa110f3de10db)

### 28th September

- [Interpreter ast_value to property value added](https://github.com/ftd-lang/ftd/commit/624ad3d37a389253ae74a5dfd12872c88023d8fd)

### 26th September

- [Interpreter Record test added](https://github.com/FifthTry/ftd/commit/ef154ac7457d6f770eb80da407505884c1b235ff)
- [Interpreter Record added](https://github.com/FifthTry/ftd/commit/acd3e986417802f467db3fdead2bd84939eb46e5)

### 25th September

- [Added condition in component property test](https://github.com/FifthTry/ftd/commit/7aa5d99db9d44f128434e0b9d2353dfd695589ca)
- [Added condition in headers test](https://github.com/FifthTry/ftd/commit/fe5170c2e1b23ec64ea7ab4e0b6e06634ec1fc22)
- [Added condition in headers](https://github.com/FifthTry/ftd/commit/7dbf77b9fb6a9f1a359f454e6f5d4e4cc1187bed)
- [Ignore comments while processing AST](https://github.com/FifthTry/ftd/commit/9285db8e9ce81f520d21ea1f62f43803407c9247)
- [Added condition in component and variable invocation](https://github.com/FifthTry/ftd/commit/b935385e64fe9d73fcf1c568c5e7415ddd07695d)
- [Added componentInvocation test](https://github.com/FifthTry/ftd/commit/cf8381c6e946835e2e4e0f47065882c78d0d9c57)

### 24th September

- [Added line_number in VariableInvocation](https://github.com/FifthTry/ftd/commit/41c4795f5ad88672314131c0336412abca0bf152)
- [Added line_number in variabledefinition and created struct for header](https://github.com/FifthTry/ftd/commit/6b7ab036cf8140fd1fb228fb91b73bf969d6966c)
- [Record field updated](https://github.com/FifthTry/ftd/commit/1cd377aa775b95e86e5698d3b5adc068de56cb72)
- [Added line_number in record](https://github.com/FifthTry/ftd/commit/11e8802a8d929a5f8bfc680683e9336b3cd79997)
- [Line number added in import](https://github.com/FifthTry/ftd/commit/e4d50c925a26f33e397828d41b67ff9b4f88f669)
- [Added events in Component](https://github.com/FifthTry/ftd/commit/bdbf64c7133327778832e728ef0839823df789e0)
- [Added iteration in Component](https://github.com/FifthTry/ftd/commit/4e839f795ddad5ab0ff008e520f54cecb96d2fe1)

### 23rd September

- [Component definition test added in AST](https://github.com/FifthTry/ftd/commit/0f2bb2a937fb2ac53760381859c8d73ac14e3d61)
- [Component Invocation added in AST](https://github.com/FifthTry/ftd/commit/e38215adfc116010b8ac20012260c786f6050242)
- [Component Definition added in AST](https://github.com/FifthTry/ftd/commit/c33dab69d5f87af4f7bb8ab865a9f3116c2989c8)

### 22nd September

- [Some minor fixes related to ast variable](https://github.com/FifthTry/ftd/commit/98acb9a0ae9dc5d04c74871326fe56a5c26070c3)
- [Added variable invocation ast](https://github.com/FifthTry/ftd/commit/3e904eb63b1a3814a0c45e1057ba4c5bf901fa03)
- [Added more test for AST::VariableDefinition](https://github.com/FifthTry/ftd/commit/9b88beee12bd62fd8087a5eca8d7c208b1b095b9)
- [Variable value done](https://github.com/FifthTry/ftd/commit/9771c08b54e4c596b44e8656d78ff077823cc98d)

### 21st September

- [Added ast for import and record](https://github.com/FifthTry/ftd/commit/eb6b724afe76618536691b52a97ec6f9bf9f4809)

### 15th September

- https://github.com/FifthTry/ftd/discussions/421
- https://github.com/FifthTry/ftd/discussions/385
- https://github.com/FifthTry/ftd/discussions/424

### 12th September

- [x] [Rename AST to DI (import, record)](https://github.com/FifthTry/ftd/commit/f21a7134be633d203a4f1739cf9446bcb5de002e)
- [x] [Added Definition in DI](https://github.com/FifthTry/ftd/commit/f21a7134be633d203a4f1739cf9446bcb5de002e)
- [x] [Added Invocation in DI](https://github.com/FifthTry/ftd/commit/2931c7ef849128a5f65b6705a332d94e14898c72)
- [x] Tests

### 11th September

Read About
- [HM type system](https://en.wikipedia.org/wiki/Hindley%E2%80%93Milner_type_system)
- [Chomsky Hierarchy in Theory of Computation](https://www.geeksforgeeks.org/chomsky-hierarchy-in-theory-of-computation/)
- [Context-free grammar](https://en.wikipedia.org/wiki/Context-free_grammar) / [Context-free language](https://en.wikipedia.org/wiki/Context-free_language)
- [Regular grammar](https://www.geeksforgeeks.org/regular-grammar-model-regular-grammars/#:~:text=Regular%20grammar%20generates%20regular%20language,T*%20i.e.%20string%20of%20terminals.)
- [Pushdown automata](https://www.geeksforgeeks.org/theory-of-computation-pushdown-automata/)
- [Finite Automata](https://www.geeksforgeeks.org/introduction-of-finite-automata/)

### 10th September

- [x] [FTD AST for import](https://github.com/FifthTry/ftd/commit/bb22aaa354bcab0e725c80c8e58f43e0e93ba117)
- [x] [FTD AST for record](https://github.com/FifthTry/ftd/commit/c8912b38de227c1d2c52360373300e4c96ae131a)

### 9th September

- [x] [Create Interpreter States (works only for the variable for now)](https://github.com/FifthTry/ftd/pull/414)
- [x] [Integrate Variable in the interpreter](https://github.com/FifthTry/ftd/pull/414)

### 8th September

- [ ] Create Interpreter States (works only for the variable for now)
- [ ] Integrate Variable in the interpreter
- [x] Implement Record

### 7th September

- [x] [Implement Variable](https://github.com/FifthTry/ftd/pull/406)

### 6th September

- [x] [Implement Variable Value/Property Value](https://github.com/FifthTry/ftd/pull/406)

### 3rd September

- [x] [Variable Kind Data](https://github.com/FifthTry/ftd/pull/404)

### 2nd September

- [x] [P1 parser](https://github.com/FifthTry/ftd/pull/401)

### 1st September

- [ ] [P1 parser](https://github.com/FifthTry/ftd/pull/401)

----
</details>

### Related Discussions

- https://github.com/FifthTry/ftd/discussions/417
- https://github.com/FifthTry/ftd/discussions/419
- https://github.com/FifthTry/ftd/discussions/413
- https://github.com/FifthTry/ftd/discussions/388
- https://github.com/FifthTry/ftd/discussions/390
- https://github.com/FifthTry/ftd/discussions/421
- https://github.com/FifthTry/ftd/discussions/381
- https://github.com/FifthTry/ftd/discussions/385
- https://github.com/FifthTry/ftd/discussions/424
- https://github.com/FifthTry/ftd/discussions/380

Say if I have to write something like this:

```ftd
-- clippy.check: absurd_extreme_comparisons
group: correctness
level: Deny
what-it-does:
> Checks for comparisons where one side of the relation is
> either the minimum or maximum value for its type and warns if it involves a
> case that is always true or always false. Only integer and boolean types are
> checked.
why-it-is-bad:
> An expression like `min <= x` may misleadingly imply
> that is is possible for `x` to be less than the minimum. Expressions like
> `max < x` are probably mistakes.
known-problems:
> For `usize` the size of the current compile target will
> be assumed (e.g., 64 bits on 64 bit systems). This means code that uses such
> a comparison to detect target pointer width will trigger this lint. 
> 
> One can use `mem::sizeof` and compare its value or conditional compilation
> attributes like `#[cfg(target_pointer_width = \"64\")] ..` instead."
example: 
> -- ft.code:
> lang: rs
> 
> let vec: Vec<isize> = vec![];
> if vec.len() <= 0 {}
> if 100 > std::i32::MAX {}
```

Currently there can be no empty lines. What if we can also write:

```ftd
-- clippy.check: absurd_extreme_comparisons
group: correctness
level: Deny

what-it-does:
> Checks for comparisons where one side of the relation is
> either the minimum or maximum value for its type and warns if it involves a
> case that is always true or always false. Only integer and boolean types are
> checked.

why-it-is-bad:
> An expression like `min <= x` may misleadingly imply
> that is is possible for `x` to be less than the minimum. Expressions like
> `max < x` are probably mistakes.

known-problems:
> For `usize` the size of the current compile target will
> be assumed (e.g., 64 bits on 64 bit systems). This means code that uses such
> a comparison to detect target pointer width will trigger this lint. 

> One can use `mem::sizeof` and compare its value or conditional compilation
> attributes like `#[cfg(target_pointer_width = \"64\")] ..` instead."

example: 
> -- ft.code:
> lang: rs

> let vec: Vec<isize> = vec![];
> if vec.len() <= 0 {}
> if 100 > std::i32::MAX {}
```

Notice the blank lines added. They improve readability. They make our body parser difficult by assuming body can not start with a line containing `:`. 

If we do indeed have more intelligent body parser, can we not do:

```ftd
-- clippy.check: absurd_extreme_comparisons
group: correctness
level: Deny

what-it-does:

Checks for comparisons where one side of the relation is
either the minimum or maximum value for its type and warns if it involves a
case that is always true or always false. Only integer and boolean types are
checked.

why-it-is-bad:

An expression like `min <= x` may misleadingly imply
that is is possible for `x` to be less than the minimum. Expressions like
`max < x` are probably mistakes.

known-problems:

For `usize` the size of the current compile target will
be assumed (e.g., 64 bits on 64 bit systems). This means code that uses such
a comparison to detect target pointer width will trigger this lint. 

One can use `mem::sizeof` and compare its value or conditional compilation
attributes like `#[cfg(target_pointer_width = \"64\")] ..` instead."
```

How do we handle the code block?

```ftd
example: 
> -- ft.code:
> lang: rs

> let vec: Vec<isize> = vec![];
> if vec.len() <= 0 {}
> if 100 > std::i32::MAX {}
```

One option is we mix multiline headers with self closing tags:

```ftd
-- clippy.check: absurd_extreme_comparisons
group: correctness
level: Deny

-- what-it-does:

Checks for comparisons where one side of the relation is
either the minimum or maximum value for its type and warns if it involves a
case that is always true or always false. Only integer and boolean types are
checked.

-- why-it-is-bad:

An expression like `min <= x` may misleadingly imply
that is is possible for `x` to be less than the minimum. Expressions like
`max < x` are probably mistakes.

-- known-problems:

For `usize` the size of the current compile target will
be assumed (e.g., 64 bits on 64 bit systems). This means code that uses such
a comparison to detect target pointer width will trigger this lint. 

One can use `mem::sizeof` and compare its value or conditional compilation
attributes like `#[cfg(target_pointer_width = \"64\")] ..` instead."

-- example: 

-- ft.code:
lang: rs

let vec: Vec<isize> = vec![];
if vec.len() <= 0 {}

if 100 > std::i32::MAX {}

-- end: ft.code

The body
```

The body is too far now. Also header fields are not namespaces, so it is not obvious if they are parameters or not. Plus it is forcing us to close the `ft.code`, which is a self closing component.

Maybe we can do this:

```ftd
-- clippy.check: absurd_extreme_comparisons
group: correctness
level: Deny

-- this.what-it-does:

Checks for comparisons where one side of the relation is
either the minimum or maximum value for its type and warns if it involves a
case that is always true or always false. Only integer and boolean types are
checked.

-- this.why-it-is-bad:

An expression like `min <= x` may misleadingly imply
that is is possible for `x` to be less than the minimum. Expressions like
`max < x` are probably mistakes.

-- this.known-problems:

For `usize` the size of the current compile target will
be assumed (e.g., 64 bits on 64 bit systems). This means code that uses such
a comparison to detect target pointer width will trigger this lint. 

One can use `mem::sizeof` and compare its value or conditional compilation
attributes like `#[cfg(target_pointer_width = \"64\")] ..` instead."

-- this.example: 

-- ft.code:
lang: rs

let vec: Vec<isize> = vec![];
if vec.len() <= 0 {}

if 100 > std::i32::MAX {}

-- this.body:

The body
```

`-- this.body` would be only needed if any `-- this.<argument-name>:` (argument continuation) has been used.

Instead of introducing `this`, we can also do:

```ftd
-- clippy.check: absurd_extreme_comparisons
group: correctness
level: Deny

-- clippy.check.what-it-does:

Checks for comparisons where one side of the relation is
either the minimum or maximum value for its type and warns if it involves a
case that is always true or always false. Only integer and boolean types are
checked.

-- clippy.check.why-it-is-bad:

An expression like `min <= x` may misleadingly imply
that is is possible for `x` to be less than the minimum. Expressions like
`max < x` are probably mistakes.

-- clippy.check.known-problems:

For `usize` the size of the current compile target will
be assumed (e.g., 64 bits on 64 bit systems). This means code that uses such
a comparison to detect target pointer width will trigger this lint. 

One can use `mem::sizeof` and compare its value or conditional compilation
attributes like `#[cfg(target_pointer_width = \"64\")] ..` instead."

-- clippy.check.example:  

;; don't have to close `.example` because example's type is `ftd.ui`
;; and depending on the actual UI constructor passed, we have to close
;; or not close. Here we are passing `ftd.code`, which is not a container
;; so it nothing has to be closed. Had we passed a `ftd.row` to `example`
;; we would had to close the `ftd.row`

-- ft.code:
lang: rs

let vec: Vec<isize> = vec![];
if vec.len() <= 0 {}

if 100 > std::i32::MAX {}

-- clippy.check.body:

The body
```

**Comment by Arpita-Jaiswal** [comment]
Currently, for caption, we use continuation syntax `>`. For instance
```ftd
-- title: 
> Hello
> This is wonderful
author: Abrar
```

What if we follow proposed syntax and write something like

```ftd
-- title:
author: Abrar

-- title.caption: 

Hello
This is wonderful
```

**Reply by Arpita-Jaiswal** [reply]
@amitu Please take a look.

**Reply by amitu** [reply]
Looks good.

**Comment by Arpita-Jaiswal** [comment]
For this 
```ftd
-- page:
header: 
> -- foo:
> bar:
>> -- ftd.text: Hello
 

-- ftd.column page:
ftd.ui header:

-- ftd.column foo:
ftd.ui bar:
```

Now, we need to do this
```ftd
-- page:

-- page.header:
  
    -- foo:
    
    -- foo.bar: 
      
          -- foo:
          
          -- this.bar: 
              
                  -- ftd.text: Hello
```

With indentation, this looks fine, but without it, this looks horrible.

**Reply by amitu** [reply]
Yep. Indentation would make things more readable in some cases. Anyways we do support indentation so its alright.

**Reply by amitu** [reply]
Also I updated the description with a syntax without `this`.

```ftd
-- page:

-- page.header:

-- foo:    
   -- foo.bar: 

        -- foo:
             -- this.bar: 
                  -- ftd.text: Hello
          -- end: foo

-- end: page.header
```

Am stumped here.

**Reply by amitu** [reply]
I think the resolution here is to force an `-- end` even for argument type `ftd.ui` where it is strictly speaking not needed as it takes a single component, and that component can chose to end itself or not depending on if it is container or not.

**Reply by amitu** [reply]
```
-- page:

-- page.header:
  
-- foo:
  -- foo.bar: 
      -- foo:
         -- foo.bar: 
           -- ftd.text: Hello
         -- end: foo.bar
  -- end: foo.bar


-- end: page.header
```

**Reply by amitu** [reply]
The key thing is we are still trying to reduce the number of ends needed, while still being unambiguous (at least to the program).

**Comment by amitu** [comment]
We can use [anonymous list](https://github.com/FifthTry/ftd/discussions/385) also without `> `s:

```ftd
-- person a:

-- person.jobs:

-- job: Intern
company: FifthTry
started: 2022-08-22
ended: 2022-12-07
 
Implemented some cool stuff.

-- job: Dev Ops 1
company: FifthTry
started: 2022-12-08

Deving some ops.

-- end: person.jobs ;; we have to end because this is a list

-- person.tags:

-- string: whatever
-- string: whatever-else

-- end: person.tags
```

**Comment by Arpita-Jaiswal** [comment]
If we have to use "continuation" during component definition, eg:

```ftd
-- ftd.row foo:
string details:
> This is foo
> Details
ftd.ui title:
> -- ftd.text: Hello
```
How would we do it?

```ftd
-- ftd.row foo:

-- string foo.details:
This is foo
Details
```

?

**Comment by Arpita-Jaiswal** [comment]
Two open points as of now:

- `ftd.ui list children` where `children` is the special keyword for container type component
- We all agree that `$` is just for referring to a variable. Using `$` to define a variable as mutable, like `-- string $name: foo` is inconsistent. Should we have a different keyword or special character or header or something instead? Shobhit has suggested using `mut` keyword.
